From 5cd6657242b7d67fb3cf0a3fa91aed9492664d1e Mon Sep 17 00:00:00 2001
From: Matt Ellis <matell@microsoft.com>
Date: Thu, 16 Mar 2017 19:20:35 -0700
Subject: [PATCH 1/3] Support building on *nix

 - Add build.sh and other required scripts
 - Move off of nuproj for describing packages, instead use nupsecs
   directly so we can use the same files when packing with `dotnet`
 - Add projects to be able to build nupkg's with `dotnet` tooling

There were some workarounds for issues with previous versions of the
tooling that are no longer problems.
---
 build.dotnet.proj                                  | 18 ++++++
 build.sh                                           | 13 ++++
 .../Microsoft.NET.Sdk.Publish.Core.nuspec          | 30 ++++++++++
 ...Microsoft.NET.Sdk.Publish.Package.DotNet.csproj | 19 ++++++
 .../Microsoft.NET.Sdk.Publish.Package.csproj       |  4 +-
 .../Microsoft.NET.Sdk.Publish.nuproj               | 70 ++--------------------
 .../Microsoft.NET.Sdk.Publish.nuspec               | 21 ++++++-
 .../Microsoft.NET.Sdk.Publish.Tasks.csproj         | 24 ++++++--
 .../Microsoft.NET.Sdk.Web.Package.DotNet.csproj    | 19 ++++++
 .../Microsoft.NET.Sdk.Web.nuproj                   | 18 ++----
 .../Microsoft.NET.Sdk.Web.nuspec                   |  4 ++
 ...NET.Sdk.Web.ProjectSystem.Package.DotNet.csproj | 19 ++++++
 .../Microsoft.NET.Sdk.Web.ProjectSystem.nuproj     | 24 ++------
 .../Microsoft.NET.Sdk.Web.ProjectSystem.nuspec     |  8 ++-
 tools/Build.sh                                     | 13 ++++
 tools/EnsureWebSdkEnv.sh                           |  6 ++
 tools/WebSdkEnv.cmd                                | 12 ++--
 tools/WebSdkEnv.sh                                 | 12 ++++
 18 files changed, 220 insertions(+), 114 deletions(-)
 create mode 100644 build.dotnet.proj
 create mode 100755 build.sh
 create mode 100644 src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.Core.nuspec
 create mode 100644 src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.Package.DotNet.csproj
 create mode 100644 src/Web/Microsoft.NET.Sdk.Web.Package/Microsoft.NET.Sdk.Web.Package.DotNet.csproj
 create mode 100644 src/Web/Microsoft.NET.Sdk.Web.ProjectSystem.Package/Microsoft.NET.Sdk.Web.ProjectSystem.Package.DotNet.csproj
 create mode 100755 tools/Build.sh
 create mode 100755 tools/EnsureWebSdkEnv.sh
 create mode 100755 tools/WebSdkEnv.sh

diff --git a/build.dotnet.proj b/build.dotnet.proj
new file mode 100644
index 0000000..186cc14
--- /dev/null
+++ b/build.dotnet.proj
@@ -0,0 +1,18 @@
+<Project InitialTargets="ValidateParameters">
+
+  <ItemGroup>
+    <ProjectFiles Include="src\Publish\Microsoft.NET.Sdk.Publish.Package\Microsoft.NET.Sdk.Publish.Package.DotNet.csproj" />
+    <ProjectFiles Include="src\Web\Microsoft.NET.Sdk.Web.ProjectSystem.Package\Microsoft.NET.Sdk.Web.ProjectSystem.Package.DotNet.csproj" />
+    <ProjectFiles Include="src\Web\Microsoft.NET.Sdk.Web.Package\Microsoft.NET.Sdk.Web.Package.DotNet.csproj" />
+  </ItemGroup>
+
+  <Target Name="ValidateParameters">
+    <Error Text="Configuration must be 'Debug' or 'Release'" Condition="'$(Configuration)' != 'Debug' and '$(Configuration)' != 'Release'" />
+    <Error Text="WebSdkVersion must be defined" Condition="'$(WebSdkVersion)' == ''" />
+    <Error Text="CoreSdkVersion must be defined" Condition="'$(CoreSdkVersion)' == ''" />
+  </Target>
+
+  <Target Name="Pack">
+    <MSBuild Projects="@(ProjectFiles)" Targets="Pack" BuildInParallel="true" />
+  </Target>
+</Project>
diff --git a/build.sh b/build.sh
new file mode 100755
index 0000000..2ccb664
--- /dev/null
+++ b/build.sh
@@ -0,0 +1,13 @@
+#!/bin/bash
+
+set -euo pipefail
+IFS=$'\n\t'
+
+SCRIPT_ROOT="$(dirname "${BASH_SOURCE[0]}")"
+DOTNET_TOOL_ROOT="$SCRIPT_ROOT/obj/tools/dotnet"
+
+curl -L https://raw.githubusercontent.com/dotnet/cli/rel/1.0.0/scripts/obtain/dotnet-install.sh | bash -s -- --version 1.0.0 --install-dir "$DOTNET_TOOL_ROOT"
+
+export PATH=$DOTNET_TOOL_ROOT:$PATH
+
+$SCRIPT_ROOT/tools/Build.sh
diff --git a/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.Core.nuspec b/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.Core.nuspec
new file mode 100644
index 0000000..5a03f82
--- /dev/null
+++ b/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.Core.nuspec
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
+<package xmlns="http://schemas.microsoft.com/packaging/2011/08/nuspec.xsd">
+  <metadata>
+    <id>Microsoft.NET.Sdk.Publish</id>
+    <version>$WebSdkVersion$</version>
+    <title>Microsoft.NET.Sdk.Publish</title>
+    <authors>Microsoft Corporation</authors>
+    <owners>Microsoft</owners>
+    <licenseUrl>https://www.microsoft.com/web/webpi/eula/net_library_eula_enu.htm</licenseUrl>
+    <projectUrl>https://github.com/aspnet/websdk</projectUrl>
+    <requireLicenseAcceptance>true</requireLicenseAcceptance>
+    <description>This package contains MSBuild targets and tasks for publishing ASP.NET projects. This package is required for publishing ASP.NET Core projects.</description>
+    <copyright>&#169; Microsoft Corporation. All rights reserved.</copyright>
+    <tags></tags>
+  </metadata>
+  <files>
+    <file src="$WebSdkRoot$bin\$Configuration$\netstandard1.3\Microsoft.NET.Sdk.Publish.Tasks.dll" target="tools\netcoreapp1.0\" />
+    <file src="$WebSdkRoot$bin\$Configuration$\netstandard1.3\Newtonsoft.Json.dll" target="tools\netcoreapp1.0\" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\ComputeTargets\*.*" target="build\netstandard1.0\ComputeTargets\" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\PublishTargets\*.*" target="build\netstandard1.0\PublishTargets\" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\CopyTargets\*.*" target="build\netstandard1.0\CopyTargets\" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\TransformTargets\*.*" target="build\netstandard1.0\TransformTargets\" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\PublishProfiles\*.*" target="build\netstandard1.0\PublishProfiles\" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\DotNetCLIToolTargets\*.*" target="build\netstandard1.0\DotNetCLIToolTargets\" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\Microsoft.NET.Sdk.Publish.props" target="build\netstandard1.0\" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\Microsoft.NET.Sdk.Publish.targets" target="build\netstandard1.0\" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\Sdk.props" target="Sdk\" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\Sdk.targets" target="Sdk\" />
+  </files>
+</package>
diff --git a/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.Package.DotNet.csproj b/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.Package.DotNet.csproj
new file mode 100644
index 0000000..be193f8
--- /dev/null
+++ b/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.Package.DotNet.csproj
@@ -0,0 +1,19 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <DisableImplicitFrameworkReferences>true</DisableImplicitFrameworkReferences>
+    <IncludeBuildOutput>false</IncludeBuildOutput>
+    <NoBuild>true</NoBuild>
+  </PropertyGroup>
+  <PropertyGroup>
+    <WebSdkRoot Condition="'$(WebSdkRoot)' == ''">$(MSBuildThisFileDirectory)..\..\..\</WebSdkRoot>
+    <WebSdkSource Condition="'$(WebSdkSource)' == ''">$(WebSdkRoot)src\</WebSdkSource>
+    <NuspecFile>Microsoft.NET.Sdk.Publish.Core.nuspec</NuspecFile>
+    <OutputPath>$(WebSdkRootDir)bin\$(Configuration)\</OutputPath>
+    <IntermediateOutputPath>$(WebSdkRootDir)obj\$(Configuration)\</IntermediateOutputPath>
+    <CoreSdkVersion Condition="'$(CoreSdkVersion)' ==''">1.0.0</CoreSdkVersion>
+    <WebSdkVersion Condition="'$(WebSdkVersion)' ==''">$(CoreSdkVersion)</WebSdkVersion>
+    <NuSpecProperties>CoreSdkVersion=$(CoreSdkVersion);WebSdkVersion=$(WebSdkVersion);WebSdkSource=$(WebSdkSource);WebSdkRoot=$(WebSdkRoot);Configuration=$(Configuration)</NuSpecProperties>
+    <NoPackageAnalysis>true</NoPackageAnalysis>
+    <PackageOutputPath>$(WebSdkRoot)bin/$(Configuration)/NuGet</PackageOutputPath>
+  </PropertyGroup>
+</Project>
diff --git a/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.Package.csproj b/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.Package.csproj
index f2eb993..93faa2b 100644
--- a/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.Package.csproj
+++ b/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.Package.csproj
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003" ToolsVersion="4.0">
   <PropertyGroup>
     <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
@@ -29,4 +29,4 @@
     <MSBuild Projects="@(NuGetProject)" Targets="Clean" />
   </Target>
   <Target Name="Rebuild" DependsOnTargets="Clean;Build" />
-</Project>
\ No newline at end of file
+</Project>
diff --git a/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.nuproj b/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.nuproj
index b3a38ea..5e588b6 100644
--- a/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.nuproj
+++ b/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.nuproj
@@ -1,77 +1,17 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003" ToolsVersion="4.0">
   <PropertyGroup>
-    
-    <WebSdkRoot Condition="'$(WebSdkRoot)' == ''">..\..\..\</WebSdkRoot>
+
+    <WebSdkRoot Condition="'$(WebSdkRoot)' == ''">$(MSBuildThisFileDirectory)..\..\..</WebSdkRoot>
     <WebSdkSource Condition="'$(WebSdkSource)' == '' ">$(WebSdkRoot)\src\</WebSdkSource>
     <WebSdkTools Condition="'$(WebSdkTools)' == '' ">$(WebSdkRoot)\tools\</WebSdkTools>
-    
+
     <NuSpecFile>$(MSBuildProjectName).nuspec</NuSpecFile>
     <OutputPath>$(WebSdkRoot)\bin\$(Configuration)\</OutputPath>
     <IntermediateOutputPath>$(WebSdkRoot)\obj\$(Configuration)\</IntermediateOutputPath>
-    <RuntimeIdenfier Condition="'$(RuntimeIdenfier)' == ''">win7-x86</RuntimeIdenfier>
     <CoreSdkVersion Condition="'$(CoreSdkVersion)' ==''">1.0.0</CoreSdkVersion>
     <WebSdkVersion Condition="'$(WebSdkVersion)' ==''">$(CoreSdkVersion)</WebSdkVersion>
-    <NuSpecProperties>CoreSdkVersion=$(CoreSdkVersion);WebSdkVersion=$(WebSdkVersion)</NuSpecProperties>
+    <NuSpecProperties>CoreSdkVersion=$(CoreSdkVersion);WebSdkVersion=$(WebSdkVersion);WebSdkRoot=$(WebSdkRoot);WebSdkSource=$(WebSdkSource);WebSdkTools=$(WebSdkTools)</NuSpecProperties>
   </PropertyGroup>
-  <ItemGroup>
-    <NuGetContent Include="$(WebSdkRoot)\bin\$(Configuration)\net46\$(RuntimeIdenfier)\Microsoft.NET.Sdk.Publish.Tasks.dll" >
-      <Destination>tools\net46</Destination>
-    </NuGetContent>
-    <NuGetContent Include="$(WebSdkRoot)\bin\$(Configuration)\net46\$(RuntimeIdenfier)\Newtonsoft.Json.dll" >
-      <Destination>tools\net46</Destination>
-    </NuGetContent>
-    <NuGetContent Include="$(WebSdkTools)\WebDeploy\**\*.*" >
-      <Destination>tools\net46\%(RecursiveDir)%(Filename)%(Extension)</Destination>
-    </NuGetContent>
-
-    <NuGetContent Include="$(WebSdkRoot)\bin\$(Configuration)\netstandard1.3\Microsoft.NET.Sdk.Publish.Tasks.dll" >
-      <Destination>tools\netcoreapp1.0</Destination>
-    </NuGetContent>
-    <NuGetContent Include="$(WebSdkRoot)\bin\$(Configuration)\net46\$(RuntimeIdenfier)\Newtonsoft.Json.dll" >
-      <Destination>tools\netcoreapp1.0</Destination>
-    </NuGetContent>
-    
-    <NuGetContent Include="$(WebSdkSource)\Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\ComputeTargets\*.*">
-      <Destination>build\netstandard1.0\ComputeTargets</Destination>
-    </NuGetContent>
-
-    <NuGetContent Include="$(WebSdkSource)\Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\PublishTargets\*.*">
-      <Destination>build\netstandard1.0\PublishTargets</Destination>
-    </NuGetContent>
-
-    <NuGetContent Include="$(WebSdkSource)\Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\CopyTargets\*.*">
-      <Destination>build\netstandard1.0\CopyTargets</Destination>
-    </NuGetContent>
-
-    <NuGetContent Include="$(WebSdkSource)\Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\TransformTargets\*.*">
-      <Destination>build\netstandard1.0\TransformTargets</Destination>
-    </NuGetContent>
-
-    <NuGetContent Include="$(WebSdkSource)\Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\PublishProfiles\*.*">
-      <Destination>build\netstandard1.0\PublishProfiles</Destination>
-    </NuGetContent>
-
-    <NuGetContent Include="$(WebSdkSource)\Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\DotNetCLIToolTargets\*.*">
-      <Destination>build\netstandard1.0\DotNetCLIToolTargets</Destination>
-    </NuGetContent>
-
-    <NuGetContent Include="$(WebSdkSource)\Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\Microsoft.NET.Sdk.Publish.props">
-      <Destination>build\netstandard1.0</Destination>
-    </NuGetContent>
-    
-    <NuGetContent Include="$(WebSdkSource)\Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\Microsoft.NET.Sdk.Publish.targets">
-      <Destination>build\netstandard1.0</Destination>
-    </NuGetContent>
-
-    <NuGetContent Include="$(WebSdkSource)\Publish\Microsoft.NET.Sdk.Publish.Targets\Sdk.props">
-      <Destination>Sdk</Destination>
-    </NuGetContent>
-    
-    <NuGetContent Include="$(WebSdkSource)\Publish\Microsoft.NET.Sdk.Publish.Targets\Sdk.targets">
-      <Destination>Sdk</Destination>
-    </NuGetContent>
-    
-  </ItemGroup>
   <Import Project="$(WebSdkRoot)\tools\NuGet\NuGetProj.targets"/>
-</Project>
\ No newline at end of file
+</Project>
diff --git a/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.nuspec b/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.nuspec
index 5657342..0901b9d 100644
--- a/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.nuspec
+++ b/src/Publish/Microsoft.NET.Sdk.Publish.Package/Microsoft.NET.Sdk.Publish.nuspec
@@ -1,4 +1,4 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?>
 <package xmlns="http://schemas.microsoft.com/packaging/2011/08/nuspec.xsd">
   <metadata>
     <id>Microsoft.NET.Sdk.Publish</id>
@@ -13,4 +13,21 @@
     <copyright>&#169; Microsoft Corporation. All rights reserved.</copyright>
     <tags></tags>
   </metadata>
-</package>
\ No newline at end of file
+  <files>
+    <file src="$WebSdkRoot$bin\$Configuration$\net46\Microsoft.NET.Sdk.Publish.Tasks.dll" target="tools\net46" />
+    <file src="$WebSdkRoot$bin\$Configuration$\net46\Newtonsoft.Json.dll" target="tools\net46" />
+    <file src="$WebSdkTools$WebDeploy\**\*.*" target="tools\net46" />
+    <file src="$WebSdkRoot$bin\$Configuration$\netstandard1.3\Microsoft.NET.Sdk.Publish.Tasks.dll" target="tools\netcoreapp1.0" />
+    <file src="$WebSdkRoot$bin\$Configuration$\netstandard1.3\Newtonsoft.Json.dll" target="tools\netcoreapp1.0" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\ComputeTargets\*.*" target="build\netstandard1.0\ComputeTargets" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\PublishTargets\*.*" target="build\netstandard1.0\PublishTargets" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\CopyTargets\*.*" target="build\netstandard1.0\CopyTargets" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\TransformTargets\*.*" target="build\netstandard1.0\TransformTargets" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\PublishProfiles\*.*" target="build\netstandard1.0\PublishProfiles" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\DotNetCLIToolTargets\*.*" target="build\netstandard1.0\DotNetCLIToolTargets" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\Microsoft.NET.Sdk.Publish.props" target="build\netstandard1.0" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\netstandard1.0\Microsoft.NET.Sdk.Publish.targets" target="build\netstandard1.0" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\Sdk.props" target="Sdk" />
+    <file src="$WebSdkSource$Publish\Microsoft.NET.Sdk.Publish.Targets\Sdk.targets" target="Sdk" />
+  </files>
+</package>
diff --git a/src/Publish/Microsoft.NET.Sdk.Publish.Tasks/Microsoft.NET.Sdk.Publish.Tasks.csproj b/src/Publish/Microsoft.NET.Sdk.Publish.Tasks/Microsoft.NET.Sdk.Publish.Tasks.csproj
index 3ebde01..3e93357 100644
--- a/src/Publish/Microsoft.NET.Sdk.Publish.Tasks/Microsoft.NET.Sdk.Publish.Tasks.csproj
+++ b/src/Publish/Microsoft.NET.Sdk.Publish.Tasks/Microsoft.NET.Sdk.Publish.Tasks.csproj
@@ -1,6 +1,7 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>net46;netstandard1.3</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)' == 'Windows_NT'">net46;netstandard1.3</TargetFrameworks>
+    <TargetFrameworks Condition="'$(OS)' != 'Windows_NT'">netstandard1.3</TargetFrameworks>
     <AssemblyName>Microsoft.NET.Sdk.Publish.Tasks</AssemblyName>
   </PropertyGroup>
   <PropertyGroup Condition=" '$(Configuration)' == 'Debug' ">
@@ -11,12 +12,12 @@
     <OutputPath>..\..\..\bin\Release\</OutputPath>
     <IntermediateOutputPath>..\..\..\obj\Release\</IntermediateOutputPath>
   </PropertyGroup>
-  <PropertyGroup Condition="'$(TargetFramework)' == 'net46'">
-    <RuntimeIdentifier>win7-x86</RuntimeIdentifier>
+  <PropertyGroup Condition="'$(TargetFramework)' == 'netstandard1.3'">
+    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
   </PropertyGroup>
   <ItemGroup Condition="'$(TargetFramework)' == 'netstandard1.3'">
-    <PackageReference Include="Microsoft.Build.Framework" Version="15.1.298-preview5" />
-    <PackageReference Include="Microsoft.Build.Utilities.Core" Version="15.1.298-preview5" />
+    <PackageReference Include="Microsoft.Build.Framework" Version="15.1.298-preview5" ExcludeAssets="Runtime" />
+    <PackageReference Include="Microsoft.Build.Utilities.Core" Version="15.1.298-preview5" ExcludeAssets="Runtime" />
     <PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
     <Compile Remove="Tasks\MsDeploy\*.cs" />
     <Compile Remove="MsDeploy\*.cs" />
@@ -47,4 +48,15 @@
       <LastGenOutput>Resources.Designer.cs</LastGenOutput>
     </EmbeddedResource>
   </ItemGroup>
-</Project>
\ No newline at end of file
+  <!-- Unlike building for net46, by default, Newtonsoft.Json.dll is not binplaced when we build netstandard1.3.
+       Setting CopyLocalLockFileAssemblies above causes all the assemblies from our lock file to be binplaced,
+       and this task filters out the ones we don't care about -->
+  <Target Name="FilterCopyLocal" DependsOnTargets="RunResolvePublishAssemblies" BeforeTargets="ResolveLockFileCopyLocalProjectDeps" Condition="'$(TargetFramework)' == 'netstandard1.3'">
+    <ItemGroup>
+      <_CopyLocalButNotPublished Include="@(AllCopyLocalItems)" Exclude="@(ResolvedAssembliesToPublish)" />
+      <_SystemCopyLocalItems Include="@(AllCopyLocalItems)" Condition="$([System.String]::new('%(Filename)').StartsWith('System')) or ('%(Filename)' == 'Microsoft.CSharp')" />
+      <AllCopyLocalItems Remove="@(_CopyLocalButNotPublished)" />
+      <AllCopyLocalItems Remove="@(_SystemCopyLocalItems)" />
+    </ItemGroup>
+  </Target>
+</Project>
diff --git a/src/Web/Microsoft.NET.Sdk.Web.Package/Microsoft.NET.Sdk.Web.Package.DotNet.csproj b/src/Web/Microsoft.NET.Sdk.Web.Package/Microsoft.NET.Sdk.Web.Package.DotNet.csproj
new file mode 100644
index 0000000..964677d
--- /dev/null
+++ b/src/Web/Microsoft.NET.Sdk.Web.Package/Microsoft.NET.Sdk.Web.Package.DotNet.csproj
@@ -0,0 +1,19 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <!-- We only use this csproj to pack an existing nuspec -->
+  <PropertyGroup>
+    <DisableImplicitFrameworkReferences>true</DisableImplicitFrameworkReferences>
+    <IncludeBuildOutput>false</IncludeBuildOutput>
+    <NoBuild>true</NoBuild>
+  </PropertyGroup>
+  <PropertyGroup>
+    <WebSdkRoot Condition="'$(WebSdkRoot)' == ''">$(MSBuildThisFileDirectory)..\..\..\</WebSdkRoot>
+    <WebSdkSource Condition="'$(WebSdkSource)' == ''">$(WebSdkRoot)src\</WebSdkSource>
+    <NuspecFile>Microsoft.NET.Sdk.Web.nuspec</NuspecFile>
+    <OutputPath>$(WebSdkRootDir)bin\$(Configuration)\</OutputPath>
+    <IntermediateOutputPath>$(WebSdkRootDir)obj\$(Configuration)\</IntermediateOutputPath>
+    <CoreSdkVersion Condition="'$(CoreSdkVersion)' ==''">1.0.0</CoreSdkVersion>
+    <WebSdkVersion Condition="'$(WebSdkVersion)' ==''">$(CoreSdkVersion)</WebSdkVersion>
+    <NuSpecProperties>CoreSdkVersion=$(CoreSdkVersion);WebSdkVersion=$(WebSdkVersion);WebSdkSource=$(WebSdkSource)</NuSpecProperties>
+    <PackageOutputPath>$(WebSdkRoot)bin/$(Configuration)/NuGet</PackageOutputPath>
+  </PropertyGroup>
+</Project>
diff --git a/src/Web/Microsoft.NET.Sdk.Web.Package/Microsoft.NET.Sdk.Web.nuproj b/src/Web/Microsoft.NET.Sdk.Web.Package/Microsoft.NET.Sdk.Web.nuproj
index 7b70235..919cd01 100644
--- a/src/Web/Microsoft.NET.Sdk.Web.Package/Microsoft.NET.Sdk.Web.nuproj
+++ b/src/Web/Microsoft.NET.Sdk.Web.Package/Microsoft.NET.Sdk.Web.nuproj
@@ -1,22 +1,14 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003" ToolsVersion="4.0">
   <PropertyGroup>
-    <WebSdkRootDir>..\..\..\</WebSdkRootDir>
-    <WebSourceDir>$(WebSdkRootDir)src\</WebSourceDir>
+    <WebSdkRoot Condition="'$(WebSdkRoot)' == ''">$(MSBuildThisFileDirectory)..\..\..\</WebSdkRoot>
+    <WebSdkSource Condition="'$(WebSdkSource)' == ''">$(WebSdkRoot)src\</WebSdkSource>
     <NuSpecFile>$(MSBuildProjectName).nuspec</NuSpecFile>
     <OutputPath>$(WebSdkRootDir)bin\$(Configuration)\</OutputPath>
     <IntermediateOutputPath>$(WebSdkRootDir)obj\$(Configuration)\</IntermediateOutputPath>
     <CoreSdkVersion Condition="'$(CoreSdkVersion)' ==''">1.0.0</CoreSdkVersion>
     <WebSdkVersion Condition="'$(WebSdkVersion)' ==''">$(CoreSdkVersion)</WebSdkVersion>
-    <NuSpecProperties>CoreSdkVersion=$(CoreSdkVersion);WebSdkVersion=$(WebSdkVersion)</NuSpecProperties>
+    <NuSpecProperties>CoreSdkVersion=$(CoreSdkVersion);WebSdkVersion=$(WebSdkVersion);WebSdkSource=$(WebSdkSource)</NuSpecProperties>
   </PropertyGroup>
-  <ItemGroup>
-    <NuGetContent Include="$(WebSourceDir)\Web\Microsoft.NET.Sdk.Web.Targets\Sdk.props">
-      <Destination>Sdk</Destination>
-    </NuGetContent>
-    <NuGetContent Include="$(WebSourceDir)\Web\Microsoft.NET.Sdk.Web.Targets\Sdk.targets">
-      <Destination>Sdk</Destination>
-    </NuGetContent>
-  </ItemGroup>
-  <Import Project="$(WebSdkRootDir)tools\NuGet\NuGetProj.targets"/>
-</Project>
\ No newline at end of file
+  <Import Project="$(WebSdkRoot)tools\NuGet\NuGetProj.targets"/>
+</Project>
diff --git a/src/Web/Microsoft.NET.Sdk.Web.Package/Microsoft.NET.Sdk.Web.nuspec b/src/Web/Microsoft.NET.Sdk.Web.Package/Microsoft.NET.Sdk.Web.nuspec
index f2b6c08..8154018 100644
--- a/src/Web/Microsoft.NET.Sdk.Web.Package/Microsoft.NET.Sdk.Web.nuspec
+++ b/src/Web/Microsoft.NET.Sdk.Web.Package/Microsoft.NET.Sdk.Web.nuspec
@@ -19,4 +19,8 @@
       <dependency id="Microsoft.NET.Sdk.Publish" version="$WebSdkVersion$" />
     </dependencies>
   </metadata>
+  <files>
+    <file src="$WebSdkSource$Web\Microsoft.NET.Sdk.Web.Targets\Sdk.props" target="Sdk\" />
+    <file src="$WebSdkSource$Web\Microsoft.NET.Sdk.Web.Targets\Sdk.targets" target="Sdk\" />
+  </files>
 </package>
diff --git a/src/Web/Microsoft.NET.Sdk.Web.ProjectSystem.Package/Microsoft.NET.Sdk.Web.ProjectSystem.Package.DotNet.csproj b/src/Web/Microsoft.NET.Sdk.Web.ProjectSystem.Package/Microsoft.NET.Sdk.Web.ProjectSystem.Package.DotNet.csproj
new file mode 100644
index 0000000..9b98b92
--- /dev/null
+++ b/src/Web/Microsoft.NET.Sdk.Web.ProjectSystem.Package/Microsoft.NET.Sdk.Web.ProjectSystem.Package.DotNet.csproj
@@ -0,0 +1,19 @@
+<Project Sdk="Microsoft.NET.Sdk">
+  <!-- We only use this csproj to pack an existing nuspec -->
+  <PropertyGroup>
+    <DisableImplicitFrameworkReferences>true</DisableImplicitFrameworkReferences>
+    <IncludeBuildOutput>false</IncludeBuildOutput>
+    <NoBuild>true</NoBuild>
+  </PropertyGroup>
+  <PropertyGroup>
+    <WebSdkRoot Condition="'$(WebSdkRoot)' == ''">$(MSBuildThisFileDirectory)..\..\..\</WebSdkRoot>
+    <WebSdkSource Condition="'$(WebSdkSource)' == ''">$(WebSdkRoot)src\</WebSdkSource>
+    <NuspecFile>Microsoft.NET.Sdk.Web.ProjectSystem.nuspec</NuspecFile>
+    <OutputPath>$(WebSdkRootDir)bin\$(Configuration)\</OutputPath>
+    <IntermediateOutputPath>$(WebSdkRootDir)obj\$(Configuration)\</IntermediateOutputPath>
+    <CoreSdkVersion Condition="'$(CoreSdkVersion)' ==''">1.0.0</CoreSdkVersion>
+    <WebSdkVersion Condition="'$(WebSdkVersion)' ==''">$(CoreSdkVersion)</WebSdkVersion>
+    <NuSpecProperties>CoreSdkVersion=$(CoreSdkVersion);WebSdkVersion=$(WebSdkVersion);WebSdkSource=$(WebSdkSource)</NuSpecProperties>
+    <PackageOutputPath>$(WebSdkRoot)bin/$(Configuration)/NuGet</PackageOutputPath>
+  </PropertyGroup>
+</Project>
diff --git a/src/Web/Microsoft.NET.Sdk.Web.ProjectSystem.Package/Microsoft.NET.Sdk.Web.ProjectSystem.nuproj b/src/Web/Microsoft.NET.Sdk.Web.ProjectSystem.Package/Microsoft.NET.Sdk.Web.ProjectSystem.nuproj
index f781f7d..728992e 100644
--- a/src/Web/Microsoft.NET.Sdk.Web.ProjectSystem.Package/Microsoft.NET.Sdk.Web.ProjectSystem.nuproj
+++ b/src/Web/Microsoft.NET.Sdk.Web.ProjectSystem.Package/Microsoft.NET.Sdk.Web.ProjectSystem.nuproj
@@ -1,28 +1,14 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003" ToolsVersion="4.0">
   <PropertyGroup>
-    <WebSdkRootDir>..\..\..\</WebSdkRootDir>
-    <WebSourceDir>$(WebSdkRootDir)src\</WebSourceDir>
+    <WebSdkRoot Condition="'$(WebSdkRoot)' == ''">$(MSBuildThisFileDirectory)..\..\..\</WebSdkRoot>
+    <WebSdkSource Condition="'$(WebSdkSource)' == ''">$(WebSdkRoot)src\</WebSdkSource>
     <NuSpecFile>$(MSBuildProjectName).nuspec</NuSpecFile>
     <OutputPath>$(WebSdkRootDir)bin\$(Configuration)\</OutputPath>
     <IntermediateOutputPath>$(WebSdkRootDir)obj\$(Configuration)\</IntermediateOutputPath>
     <CoreSdkVersion Condition="'$(CoreSdkVersion)' ==''">1.0.0-rc2</CoreSdkVersion>
     <WebSdkVersion Condition="'$(WebSdkVersion)' ==''">$(CoreSdkVersion)</WebSdkVersion>
-    <NuSpecProperties>CoreSdkVersion=$(CoreSdkVersion);WebSdkVersion=$(WebSdkVersion)</NuSpecProperties>
+    <NuSpecProperties>CoreSdkVersion=$(CoreSdkVersion);WebSdkVersion=$(WebSdkVersion);WebSdkSource=$(WebSdkSource)</NuSpecProperties>
   </PropertyGroup>
-  <ItemGroup>
-    <NuGetContent Include="$(WebSourceDir)\Web\Microsoft.NET.Sdk.Web.ProjectSystem.Targets\netstandard1.0\Microsoft.NET.Sdk.Web.ProjectSystem.props">
-      <Destination>build\netstandard1.0</Destination>
-    </NuGetContent>
-    <NuGetContent Include="$(WebSourceDir)\Web\Microsoft.NET.Sdk.Web.ProjectSystem.Targets\netstandard1.0\Microsoft.NET.Sdk.Web.ProjectSystem.targets">
-      <Destination>build\netstandard1.0</Destination>
-    </NuGetContent>
-    <NuGetContent Include="$(WebSourceDir)\Web\Microsoft.NET.Sdk.Web.ProjectSystem.Targets\Sdk.props">
-      <Destination>Sdk</Destination>
-    </NuGetContent>
-    <NuGetContent Include="$(WebSourceDir)\Web\Microsoft.NET.Sdk.Web.ProjectSystem.Targets\Sdk.targets">
-      <Destination>Sdk</Destination>
-    </NuGetContent>
-  </ItemGroup>
-  <Import Project="$(WebSdkRootDir)tools\NuGet\NuGetProj.targets"/>
-</Project>
\ No newline at end of file
+  <Import Project="$(WebSdkRoot)tools\NuGet\NuGetProj.targets"/>
+</Project>
diff --git a/src/Web/Microsoft.NET.Sdk.Web.ProjectSystem.Package/Microsoft.NET.Sdk.Web.ProjectSystem.nuspec b/src/Web/Microsoft.NET.Sdk.Web.ProjectSystem.Package/Microsoft.NET.Sdk.Web.ProjectSystem.nuspec
index 200689b..49b5974 100644
--- a/src/Web/Microsoft.NET.Sdk.Web.ProjectSystem.Package/Microsoft.NET.Sdk.Web.ProjectSystem.nuspec
+++ b/src/Web/Microsoft.NET.Sdk.Web.ProjectSystem.Package/Microsoft.NET.Sdk.Web.ProjectSystem.nuspec
@@ -14,4 +14,10 @@
     <projectUrl>https://github.com/aspnet/websdk</projectUrl>
     <copyright>&#169; Microsoft Corporation. All rights reserved.</copyright>
   </metadata>
-</package>
\ No newline at end of file
+  <files>
+    <file src="$WebSdkSource$Web\Microsoft.NET.Sdk.Web.ProjectSystem.Targets\netstandard1.0\Microsoft.NET.Sdk.Web.ProjectSystem.props" target="build\netstandard1.0\" />
+    <file src="$WebSdkSource$Web\Microsoft.NET.Sdk.Web.ProjectSystem.Targets\netstandard1.0\Microsoft.NET.Sdk.Web.ProjectSystem.targets" target="build\netstandard1.0\" />
+    <file src="$WebSdkSource$Web\Microsoft.NET.Sdk.Web.ProjectSystem.Targets\Sdk.props" target="Sdk\" />
+    <file src="$WebSdkSource$Web\Microsoft.NET.Sdk.Web.ProjectSystem.Targets\Sdk.targets" target="Sdk\" />
+  </files>
+</package>
diff --git a/tools/Build.sh b/tools/Build.sh
new file mode 100755
index 0000000..30ae273
--- /dev/null
+++ b/tools/Build.sh
@@ -0,0 +1,13 @@
+#!/bin/bash
+set -euo pipefail
+IFS=$'\n\t'
+
+SCRIPT_ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+
+. "$SCRIPT_ROOT/EnsureWebSdkEnv.sh"
+
+dotnet restore "$WebSdkRoot/src/Publish/Microsoft.NET.Sdk.Publish.Tasks/Microsoft.NET.Sdk.Publish.Tasks.csproj" /p:Configuration=Release
+
+dotnet build "$WebSdkRoot/src/Publish/Microsoft.NET.Sdk.Publish.Tasks/Microsoft.NET.Sdk.Publish.Tasks.csproj" /p:Configuration=Release
+
+dotnet pack "$WebSdkRoot/build.dotnet.proj" /p:Configuration=Release
diff --git a/tools/EnsureWebSdkEnv.sh b/tools/EnsureWebSdkEnv.sh
new file mode 100755
index 0000000..bc22ef7
--- /dev/null
+++ b/tools/EnsureWebSdkEnv.sh
@@ -0,0 +1,6 @@
+SCRIPT_ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"
+
+if [ -z "${WebSdkRoot:-}" ]; then
+    echo Initializing web sdk environment
+    . "$SCRIPT_ROOT/WebSdkEnv.sh"
+fi
diff --git a/tools/WebSdkEnv.cmd b/tools/WebSdkEnv.cmd
index 6226cf0..396bd43 100644
--- a/tools/WebSdkEnv.cmd
+++ b/tools/WebSdkEnv.cmd
@@ -27,12 +27,12 @@ if exist "%WebSdkProgramFiles%\Microsoft Visual Studio\VS15Preview\MSBuild\15.0\
 )
 
 set "WebSdkRoot=%~dp0"
-set "WebSdkRoot=%WebSdkRoot:~0,-7%"
-set "WebSdkBin=%WebSdkRoot%\bin\"
-set "WebSdkIntermediate=%WebSdkRoot%\obj\"
-set "WebSdkReferences=%WebSdkRoot%\references\"
-set "WebSdkSource=%WebSdkRoot%\src\"
-set "WebSdkTools=%WebSdkRoot%\tools\"
+set "WebSdkRoot=%WebSdkRoot:~0,-6%"
+set "WebSdkBin=%WebSdkRoot%bin\"
+set "WebSdkIntermediate=%WebSdkRoot%obj\"
+set "WebSdkReferences=%WebSdkRoot%references\"
+set "WebSdkSource=%WebSdkRoot%src\"
+set "WebSdkTools=%WebSdkRoot%tools\"
 
 set "PATH=%PATH%;%WebSdkMSBuildPath%"
 set "PATH=%PATH%;%WebSdkTools%"
\ No newline at end of file
diff --git a/tools/WebSdkEnv.sh b/tools/WebSdkEnv.sh
new file mode 100755
index 0000000..2ca053b
--- /dev/null
+++ b/tools/WebSdkEnv.sh
@@ -0,0 +1,12 @@
+#!/bin/bash
+
+SCRIPT_ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}" )" && pwd)"
+
+export WebSdkRoot="$(dirname $SCRIPT_ROOT)/"
+export WebSdkBin="${WebSdkRoot}bin/"
+export WebSdkIntermediate="${WebSdkRoot}obj/"
+export WebSdkReferences="${WebSdkRoot}references/"
+export WebSdkSource="${WebSdkRoot}src/"
+export WebSdkTools="${WebSdkRoot}tools/"
+
+export PATH="$WebSdkTools:$PATH"
-- 
1.8.3.1

